### PR TITLE
Use a SmallPtrSet instead of a DenseSet. More memory efficient

### PR DESCRIPTION
**Pull request by trentxintong**

### CODE REVIEW

**Comment by trentxintong** [comment]
@swift-ci Please test

File: include/swift/SILOptimizer/Utils/Local.h [hunk 1]
```diff
@@ -79,7 +79,8 @@ bool isIntermediateRelease(SILInstruction *I,
 
 /// \brief Recursively collect all the uses and transistive uses of the
 /// instruction.
-void collectUsesOfValue(SILValue V, llvm::DenseSet<SILInstruction *> &Insts);
+void
+collectUsesOfValue(SILValue V, llvm::SmallPtrSet<SILInstruction *, 4> &Insts);
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/Local.h]: IIRC there is a SmallPtrSetImpl. I think you can use it like llvm::SmallVectorImpl and thus avoid the need to put in the size here.

But my memory may be incorrect.

**Comment by trentxintong** [comment]
@swift-ci Please test and merge

File: lib/SILOptimizer/Utils/Local.cpp [hunk 2]
```diff
@@ -207,22 +207,21 @@ void swift::eraseUsesOfInstruction(SILInstruction *Inst,
...
 
     // Collect the users of this instruction.
-    Insts.insert(User);
     collectUsesOfValue(User, Insts);
   }
 }
 
 void swift::eraseUsesOfValue(SILValue V) {
-  llvm::DenseSet<SILInstruction *> Insts;
+  llvm::SmallPtrSetImpl<SILInstruction *> Insts;
```
Comment (jrose-apple) [inline lib/SILOptimizer/Utils/Local.cpp]: This should be the concrete type `llvm::SmallPtrSet`.

**Comment by trentxintong** [comment]
@swift-ci Please test and merge

**Comment by gottesmm** [comment]
@trentxintong The Impl/NonImpl pattern looks good!

**Pull request merged**